Transcold: Add detailed support.

crankyoldgit opened the pull request:
* Manually merge in the work done by @iamDshetty
* Lots of code clean up to get it working properly.
  - Code style
  - Better constants.
  - Lint
  - Bugs
  - Remove dead/unused code & settings.
* Add unit test for basic features.
* Settings/Features supported:
  - Mode
  - Fan speed
  - Power
  - Temp
  - Swing

* Sending confirmed working in: https://github.com/crankyoldgit/IRremoteESP8266/issues/1256#issuecomment-696055698

Fixes #1256

crankyoldgit committed:
Manually merge Transcold src.zip file.

Minimal alterations to the supplied code were done in this merge.
Ref: https://github.com/crankyoldgit/IRremoteESP8266/issues/1256#issuecomment-693870428

crankyoldgit committed:
Fix compiler errors.

crankyoldgit committed:
Fix all the linter errors.

crankyoldgit committed:
Use constants instead of ticks.

crankyoldgit committed:
Fix coding errors.
* Move back to the original timings.
* Adjust unit tests for new bit size.
* Add the protocol as IRac supported.
* Remove a bunch of useless debug statements.
* Add missing part of the IR message footer.
* Fix other incorrect parts of the protocol format (send & decode).

crankyoldgit committed:
Code style clean ups & optimisation.

crankyoldgit committed:
Transcold: Unit tests and remove unused/unsupported features.
* Add some unit tests to cover functionality.
* Remove turbo, light, clean, sensor, follow, sleep, etc ...
* Rename class to fit in with naming guidelines.
* Fix default/initial state.
* Other misc fixes.

crankyoldgit committed:
Fix linter issue

NiKiZe commented on src/ir_Transcold.cpp:
```diff
@@ -1,93 +1,542 @@
+    space(kDefaultMessageGap);
   }
 }
+#endif
```
Missing `  // SEND_TRANSCOLD` comment?

crankyoldgit replied:
Nice catch. Fixed.

NiKiZe commented on src/ir_Transcold.cpp:
```diff
@@ -1,93 +1,542 @@
   return true;
 }
-#endif  // DECODE_TRANSCOLD
+#endif
```
`  // DECODE_TRANSCOLD` removed?

crankyoldgit replied:
D'oh. I blindly cut-n-pasted the code. My bad. Fixed.

NiKiZe commented on src/ir_Transcold.h:
```diff
@@ -0,0 +1,187 @@
+    0b0010,  // 28C
+    0b1100,  // 29C
+    0b0100   // 30C
+};
```
Hmm, is min temp actually 18 and not 17? is there really no logic to this madness?

crankyoldgit replied:
@iamDshetty What's the story here with this comment/code? What is the actual min temp? Why is 17C the same as 18C? etc etc?

iamDshetty replied:
@crankyoldgit @NiKiZe 
Min temp of Transcold AC is 18 deg C.
We have a general UI for all AC's which starts from 17deg C. So In general 17C is required for us. So kept same as 18C as there is no 17C for this AC. 

The AC which is with me is an Old model of Transcold, doesn't mean there is no new brand of Transcold. Transcold brand is still available with new Models.

I have shared the bit calculation sheet. You can create a new file for this AC with min file size.

I will definitely help you out with Testing.

crankyoldgit replied:
I've redone how temperature is encoded/decoded. I've replaced the lookup-table with bit manipulation and math to make it safer and faster.

It now uses a min temp of 18C too.

crankyoldgit replied:
@NiKiZe You should probably re-review that change/part etc.

NiKiZe reviewed (commented):
This feels like a lot of code (flash size) for an old unit :/

crankyoldgit committed:
Responde to code review feedback.

crankyoldgit committed:
Set the min temp to 18C
Per https://github.com/crankyoldgit/IRremoteESP8266/pull/1278#discussion_r493240717

crankyoldgit committed:
Replace temp lookup tab with bit math.
* Faster & safer.
* Should be similar space-wise.

crankyoldgit commented:
> This feels like a lot of code (flash size) for an old unit :/

Well, the library supports my 20 year old Fujitsu ... so ...  :-/

This is why we allow people to tune what protocols get included/excluded etc.
